### PR TITLE
Better HTML Handling

### LINKED ISSUE #89: type() function not working in python
For some reason I cannot run the print out a data type. 

e.g. 

```run-python
print(type(35))
```

This simply does not return anything in obsidian although we would expect it to return `int`

### PR DESCRIPTION
**Pull request by chlohal**
Currently, anything printed to `stdout` is rendered as HTML. This causes problems in some languages (as per #89). 

This PR fixes that by adding a new feature to the Outputter: a toggle-HTML sigil. The Outputter will output as plaintext by default, and only swap to HTML when it sees the sigil.

In order to fix the use-case of creating dynamic HTML elements (as per #83), I added a new magic function: `@html()`. This works in JS and Python, and prints its first argument surrounded by the sigil.

`@show(...)`, Python's `plt.show`, R's `plot()`, and the new `@html(...)` function all have been altered to use the sigil.

This PR closes #89

### CODE REVIEW

File: src/Outputter.ts [hunk 1]
```diff
@@ -57,9 +61,39 @@ export class Outputter extends EventEmitter {
...
+	 * processing `toggleHtmlSigil`s along the way.
+	 * `toggleHtmlSigil`s may be interleaved with text and HTML
+	 * in any way; this method will correctly interpret them.
+	 * @param text The stdout data in question
+	 */
+	private processSigilsAndWriteText(text: string) {
+		//Loop around, removing HTML toggling sigils
+		while (true) {
+			let index = text.indexOf(TOGGLE_HTML_SIGIL);
+			if (index == -1) break;
```
Comment (milan338) [inline src/Outputter.ts:79]: Could use strict equality everywhere instead of regular equality

**Comment by milan338** [comment]
The escaping looks to work fine with `@html('string')` in python, but `@show` escapes the html in python for me, so the image doesn't show up and instead it's just the html text, and inline plots don't appear correctly:

image


Although looking into this more, it looks like python plots were broken before this pr. They work fine in master, maybe the python executor doesn't handle this properly? Should look like

image

```python
import matplotlib.pyplot as plt  
import numpy as np  

xpoints = np.array([1, 8])  
ypoints = np.array([3, 10])

plt.plot(xpoints, ypoints)  
plt.show()
```

**Comment by chlohal** [comment]
`@show(...)` fixed! I forgot to push those commits; sorry!

Oddly, `plt.show` seems to work fine for me. What version of Python are you using?
image

File: src/Outputter.ts (hunk 1 above)
Comment (chlohal) [reply src/Outputter.ts]: Fixed; thank you!

**Comment by milan338** [comment]
I'm running python 3.10, looking into this a bit more it looks like the output is being split into two separate calls to stdout which is messing it up:

image

I've just printed that out by adding this to `PythonExecutor.ts`:

```ts
const writeToStdout = (data: any) => {
    let str = data.toString();
    console.log(str);
    // ...
```

I'm not sure if this is just on my end, but pressing run again after running it the first time causes nothing to be output, and the run and clear buttons disappear. This only happens in notebook mode though.

**Comment by chlohal** [comment]
It's odd that it gets split up. Maybe that's a platform-specific issue that we have to handle w/ buffering?

Weird, I see the same issue with pressing run again. I didn't notice that-- thank you.
When you kill the runtime with the `Code Runtime Management` command, does the python process disappear?

**Comment by twibiral** [comment]
I did not experience this problem with Python 3.10 (Window 10)

**Comment by chlohal** [comment]
> I did not experience this problem with Python 3.10 (Window 10)
if i may ask, are you talking about the splitting issue, or the Notebook Mode re-running issue? 

Regarding the splitting issue, @milan338 I think that I fixed it? I'm not sure, but I believe that buffering output until it sees the next toggleSigil will fix that.

I'm trying to fix the re-running issue: it seems that, after the `plt.show()` block is ran once, stdout stops functioning correctly. I'm fairly certain it's not an issue with flushing the stream, but I'm unsure as to what it *is*.

I do know that it's an issue with matplotlib specifically. I tested a version without matplotlib, and it works fine:
image

**Comment by milan338** [comment]
The plots aren't broken anymore on my end, thanks for the fix. It looks like when the runtime is killed, the process disappears and the run code button returns and the code block can be run again as intended.

**Comment by chlohal** [comment]
>  It looks like when the runtime is killed, the process disappears and the run code button returns and the code block can be run again as intended.

Great! I was worried that was breaking too. It's wonderful to hear it's working as intended.

> I think I've found a fix by changing the matplotlib backen to a non-interactive one like 'agg'. I've submitted a PR for this, the fix works on my end. Other than that, I think this PR is ready to be merged.

It works on my end too. Thank you so much, Milan! I'm just curious, but do you know anything more about the root cause?

**Comment by milan338** [comment]
I'm still looking into what's causing the issue, I've narrowed it down to `plt.plot()` on my end, but there's more issues in #108 with showing plots that aren't embedded in the note.

**Comment by chlohal** [comment]
I also see `plt.plot()` as being the issue. 

Since Notebook Mode is relatively new, we could also say that embedding plots is mandatory in Notebook Mode and not break many workflows. Obviously, this isn't a preferred option, but we could use it as a way to get the rest of these fixes (SVG buffering, HTML-escaping, the background process blocking the runtime) out of the door so that users aren't experiencing the bugs.

If we go that route, I could push a co-authored commit by you to *this* PR so it has the `use("agg")` fix, then we could leave #108 open and make an issue to note the known problem? Open to discussion on how to get that done / if we should do it :)

**Comment by milan338** [comment]
Sounds like a good idea, thanks. I think coauthoring the fix into this PR would be best so we're actively tracking the issue in the other open PR. Maybe we shouldn't impose any restrictions right now, just leave it as an open issue and add a description to the settings that toggling embed plots off doesn't work properly with notebook mode yet?

**Comment by chlohal** [comment]
Sounds good 👍

**Comment by twibiral** [comment]
Thank you both so much! Is this PR ready to get merged?

**Comment by chlohal** [comment]
I believe it is 👍 All works fine on my end-- tested Python `@show`, `plt.show()`, and `@html`!

**Comment by twibiral** [comment]
Great!

**Comment by twibiral** [comment]
Can you add an explanation for `@html` to the Readme?

**Pull request merged**